src/render.py

```diff
@@ -68,33 +71,76 @@
 
 
 class MappedImage:
-    def __init__(self, name: str, id: int, extension: str):
+    def __init__(self, name: str, id: int):
         self.name = name
         self.id = id
-        self.ext = extension
-
-    def to_filename(self, id_width: int):
-        return "Images/%s%.*i.%s" % (self.name, id_width, self.id, self.ext)
+        self.downloaded = False
+        self.is_null = False
+        self.file = None
+        self.media_type = None
+        self.ext = None
+
+    def add_file(self, file: Optional[bytes], url: str):
+        self.downloaded = True
+        if file is None:
+            self.is_null = True
+        processed = process_image_for_epub3(file)
+        if processed is None:
+            print(
+                "Downloaded %s, but it wasn't an image of EPUB3-compatible format" % url
+            )
+            self.is_null = True
+        else:
+            self.file, self.media_type, self.ext = processed
+
+    def get_filename(self, id_width: int) -> Optional[str]:
+        if not self.downloaded:
+            raise RuntimeError(
+                "Attempted to get mapped image filename before getting it as a file. (This indicates a prior map population failure.)"
+            )
+        elif self.is_null:
+            return None
+        else:
+            return "Images/%s%.*i.%s" % (self.name, id_width, self.id, self.ext)
 
 
 class ImageMap:
     def __init__(self):
-        self.map = {}
+        self.icons = {}
+        self.images = {}
```

Comment:
We do if we want separate numbering schemes for them with correct zero-padding on each. (Which strikes me as valuable from an internal readability standpoint, to make it easier to e.g. quickly scan for one of relatively few inline images amid lots and lots of different icons.)

Reply:
I agree that the separate naming and numbering is valuable. But if I'm reading the code right, you get that by setting the name in the `MappedImage` constructor to `"icon"` or `"image"`, and incrementing the appropriate ID (which is done by `add_icon` and `add_image`). Once you've done that, it seems like you're good, and can dump both into the same map without forgetting which type of image it is.

Reply:
Good point. I'd been avoiding that on the basis that it would require string-matching to determine which retrieval function to use in `get_images_as_epub_items`, but when I looked at it a bit more that still turned out to be somewhat simpler than the version with two separate maps. The return to a single-map setup is now accordingly pushed.